src/steps/findings.ts

```diff
@@ -0,0 +1,176 @@
+  const apiClient = createAPIClient(instance.config);
+
+  await apiClient.iterateFindings(async (finding) => {
+    const findingProps = finding.resource;
+    const findingEntity = await jobState.addEntity(
+      createIntegrationEntity({
+        entityData: {
+          source: finding,
+          assign: {
+            _type: 'cobalt_finding',
+            _class: 'Finding',
+            _key: findingProps.id,
+            tag: findingProps.tag,
+            name: findingProps.title,
+            displayName: findingProps.title,
+            description: findingProps.description,
+            category: 'Penetration Testing',
+            typeCategory: findingProps.type_category,
+            labels: JSON.stringify(findingProps.labels, null, 2),
+            impact: JSON.stringify(findingProps.impact, null, 2), //required to be a string in J1 Finding
+            severity: JSON.stringify(findingProps.impact, null, 2), //required property in J1 Finding
+            numericSeverity: findingProps.impact, //required property in J1 Finding
+            likelihood: findingProps.likelihood,
+            state: findingProps.state,
+            open: true, //required property in J1 Finding
+            affectedTargets: JSON.stringify(
+              findingProps.affected_targets,
+              null,
+              2,
+            ),
+            proofOfConcept: findingProps.proof_of_concept,
+            suggestedFix: findingProps.suggested_fix,
+            prerequisites: findingProps.prerequisites,
+            pentestId: findingProps.pentest_id, //value of pentest Assessment _key
+            assetId: findingProps.asset_id, // value of asset _key (which could be class Application or something else)
+            log: JSON.stringify(findingProps.log, null, 2),
+          },
+        },
+      }),
+    );
+
+    //can't have a Finding without an Assessment (pentest)
+    const assessmentEntity = await jobState.findEntity(findingProps.pentest_id);
+    if (!assessmentEntity) {
+      throw new IntegrationMissingKeyError(
+        `Expected Assessment with key to exist (key=${findingProps.pentest_id}) as part of Finding (key=${findingProps.id})`,
+      );
+    }
+    await jobState.addRelationship(
+      createDirectRelationship({
+        _class: RelationshipClass.IDENTIFIED,
+        from: assessmentEntity,
+        to: findingEntity,
+      }),
+    );
+
+    //we would like to tie Finding to a cobalt_asset, but the asset can be deleted or not relevant
+    let assetEntity;
+    if (findingProps.asset_id) {
+      assetEntity = await jobState.findEntity(findingProps.asset_id);
+    }
+    if (assetEntity) {
+      await jobState.addRelationship(
+        createDirectRelationship({
+          _class: RelationshipClass.HAS,
+          from: assetEntity,
+          to: findingEntity,
+        }),
+      );
+    } //if assetEntity does not exist, just move on
+
+    const vulnLink: string = getVulnerabilityLink(findingProps.description);
```

Comment:
This is great!

The `Vulnerability` entities are meant to be "account global" objects, where the J1 user could search for a CVE identifier, find a single entity, and then see all related `Finding` entities, no matter the scanner / vulnerability management system. This means that we need to let the mapping system manage the life of the CVE / CWE `Vulnerability`.

The Qualys integration has code that produces the mapped relationships. Note that the `Vulnerability` would have `_type: 'CVE'`, since it is a global entity (not specific to or managed by Cobalt).

@JupiterOne/integrations @philidem here is another case where we need a CVE database management system in J1. Otherwise, this integration is, like Qualys etc., going to have to generate mapped relationships that duplicate the vulnerability data for each `Finding`.